File: TP3DPBO2425C3/CPP/KodeProgram/Dosen.cpp
#include "Pegawai.cpp" 

#ifndef DOSEN_CPP
#define DOSEN_CPP

class Dosen : public Pegawai, public IKepemimpinan, public IKehadiran {
private:
    std::string nidn;
    std::string jabatanAkademik;
    std::string jabatanStruktural;
    int jumlahSKS;
    std::string unitKerja;

public:
    Dosen(const std::string& id, const std::string& nama, double gaji, const std::string& alamat, const std::string& tglLahir,
          const std::string& nidn, const std::string& jabAkademik, const std::string& jabStruktural, int sks, const std::string& unitKerja)
        : Pegawai(id, nama, gaji, alamat, tglLahir), nidn(nidn), jabatanAkademik(jabAkademik),
          jabatanStruktural(jabStruktural), jumlahSKS(sks), unitKerja(unitKerja) {}

    // GETTER/SETTER Spesifik
    std::string getJabatanStruktural() const { return jabatanStruktural; }
    void setJabatanStruktural(const std::string& jab) { this->jabatanStruktural = jab; }

    // Implementasi Polimorfik wajib
    void tampilkanData() const override {
        tampilkanDasar();
        std::cout << "Jabatan: Dosen\n";
        std::cout << "NIDN: " << nidn << std::endl;
        std::cout << "Jabatan Akademik: " << jabatanAkademik << std::endl;
        std::cout << "Jumlah SKS Mengajar: " << jumlahSKS << std::endl;
        std::cout << "Jabatan Struktural: " << jabatanStruktural << std::endl; 
        std::cout << "Unit Kerja Struktural: " << getUnitKerja() << std::endl; 
        
        std::cout << "Keterangan Struktural: " << ambilKeputusanStrategis() << std::endl;
        std::cout << "Total Gaji + Tunjangan: " << (gajiPokok + hitungTunjangan()) << std::endl;
        std::cout << "Jam Kerja Bulan Ini: " << hitungJamKerja() << " jam\n";
    }

    double hitungTunjangan() const override {
        double tunjanganSKS = jumlahSKS * 100000; 
        double tunjanganJabatan = jabatanStruktural.empty() ? 0 : 3000000;
        return tunjanganSKS + tunjanganJabatan;
    }

    // Implementasi wajib dari IKepemimpinan (Logic Hybrid Inheritance)
    std::string ambilKeputusanStrategis() const override {
        if (!jabatanStruktural.empty()) {
            return "Mengambil keputusan strategis untuk " + getUnitKerja();
        } else {
            return "Tidak memiliki jabatan struktural aktif.";
        }
    }
    std::string getUnitKerja() const override { return unitKerja; }

    // Implementasi wajib dari IKehadiran
    double hitungJamKerja() const override { return jumlahSKS * 1.5 * 16; }
    std::string getBulanKehadiran() const override { return "Oktober"; }
};

#endif // DOSEN_CPP

File: TP3DPBO2425C3/Python/KodeProgram/Dosen.py
# Dosen.py
from Pegawai import Pegawai
from IKepemimpinan import IKepemimpinan
from IKehadiran import IKehadiran

class Dosen(Pegawai, IKepemimpinan, IKehadiran):
    """Subkelas Dosen: Hybrid Inheritance (Pegawai + IKepemimpinan)."""
    def __init__(self, id_pegawai, nama, gaji_pokok, alamat, tgl_lahir, nidn, jab_akademik, jab_struktural, sks, unit_kerja):
        super().__init__(id_pegawai, nama, gaji_pokok, alamat, tgl_lahir)
        self._nidn = nidn
        self._jabatan_akademik = jab_akademik
        self._jabatan_struktural = jab_struktural
        self._jumlah_sks = sks
        self._unit_kerja = unit_kerja

    # --- GETTER SPESIFIK ---
    @property
    def nidn(self): return self._nidn
    @property
    def jabatan_akademik(self): return self._jabatan_akademik
    @property
    def jabatan_struktural(self): return self._jabatan_struktural
    @property
    def jumlah_sks(self): return self._jumlah_sks
    
    # --- SETTER SPESIFIK ---
    @jabatan_struktural.setter
    def jabatan_struktural(self, value):
        if value not in ["Kaprodi", "Dekan", None, ""]:
            raise ValueError("Jabatan struktural tidak valid.")
        self._jabatan_struktural = value
    
    @jumlah_sks.setter
    def jumlah_sks(self, value):
        if value < 0 or value > 24: raise ValueError("SKS harus antara 0-24.")
        self._jumlah_sks = value

    def hitung_tunjangan(self):
        tunjangan_sks = self.jumlah_sks * 100000 
        tunjangan_jabatan = 3000000 if self.jabatan_struktural else 0
        return tunjangan_sks + tunjangan_jabatan

    def tampilkan_data(self):
        self._tampilkan_dasar()
        print(f"Jabatan: Dosen")
        # Menampilkan semua atribut spesifik (Pertanggungjawaban Data)
        print(f"NIDN: {self.nidn}") 
        print(f"Jabatan Akademik: {self.jabatan_akademik}")
        print(f"Jumlah SKS Mengajar: {self.jumlah_sks}")
        print(f"Jabatan Struktural: {self.jabatan_struktural}") 
        print(f"Unit Kerja Struktural: {self.get_unit_kerja()}") 
        
        # Output Hybrid Inheritance
        print(f"Keterangan Struktural: {self.ambil_keputusan_strategis()}") 
        print(f"Total Gaji + Tunjangan: {(self.gaji_pokok + self.hitung_tunjangan()):,.2f}")
        print(f"Jam Kerja Bulan Ini: {self.hitung_jam_kerja()} jam")


    # Implementasi Wajib dari IKepemimpinan (Logic Hybrid Inheritance & Enkapsulasi)
    def ambil_keputusan_strategis(self):
        """Memastikan output kepemimpinan hanya muncul jika ada jabatan."""
        # Mengakses atribut melalui Getter
        if self.jabatan_struktural: 
            return f"Mengambil keputusan strategis untuk {self.get_unit_kerja()}"
        else:
            # Output jika jabatan kosong (Pertanggungjawaban data kosong)
            return "Tidak memiliki jabatan struktural aktif." 
            
    def get_unit_kerja(self):
        # Mengembalikan field private
        return self._unit_kerja 

    # Implementasi Wajib dari IKehadiran
    def hitung_jam_kerja(self):
        return self.jumlah_sks * 1.5 * 16

    def get_bulan_kehadiran(self):
        return "Oktober"